### ISSUE TITLE
Auto device placement for distributed runtime

### ISSUE THREAD

**Issue by windreamer** [tensorflow/tensorflow#2126]
In a distributed TF setting, we need to place variables and ops to different devices. This is annoying to manually assign each variable and op, especially when we have GPU resources in our environment.

TF offer a context named `tf.train.replica_device_setter` which place variables to ps devices in round-robin manner, and this is helpful but not enough. @mrry can you shed some light on the auto distributed devices placement problem?

**Comment by windreamer** [comment]
@bhack this is the best I can do for now, pls feel free to correct me and provide some more thoughts. thx.

**Comment by bhack** [comment]
Thank you @windreamer. I also  find  this topic quite crucial for the "user experience" when we have distributed devices. Generally on cluster we could have really unbalanced GPU/CPU scalar resources ratio and there are different consideration to do for data parallel and model parallel approach. So it is really interesting to have some feedbacks. /cc @mtamburrano @lenlen

**Comment by mrry** [comment]
@vrv is working on improving the device placement so that policies will be more declarative, rather than binding to specific devices at graph construction time. This should address some of the limitations of `tf.train.replica_device_setter()`. If you had specific requirements, it would be good to collect them here, so that we could take them into consideration in the redesign.

**Comment by ZhuFengdaaa** [comment]
I'm facing [the bug here](https://github.com/tensorflow/models/issues/56). I'm wondering if it is caused by the `tf.train.replica_device_setter`. If not, what is the most possible reason should it be caused ? *Thank you*.

**Comment by mrry** [comment]
See the discussion on that issue. The problem is most likely caused by the fact that `tf.train.SyncReplicasOptimizer` bakes the number of workers into the checkpoint, which is another issue that should probably be addressed.

**Comment by vrv** [comment]
Indeed, this is a much bigger project that we'll chip away at over time (optimizing for runtime across a distributed cluster is kind of a hard problem, and we're not good enough at it that manual placement typically produces better results).  We'd love to do a better job of this, so we'll be looking at this over the next little while.  Happy to brainstorm ideas here.

**Comment by suiyuan2009** [comment]
It would be better if parameter can be partitioned into pieces and well managed.

**Comment by drpngx** [comment]
Closing due to inactivity. Feel free to open a new github issue if the problem still persists in recent versions.

**Comment by bhack** [comment]
@drpngx From last @vrv comment we are waiting news from google on this. Any progress or public roadmap?

**Comment by vrv** [comment]
@benoitsteiner is working on better cost models for auto-assignment I think, but I don't know what the expected roadmap for that work is.

C++ shape inference theoretically makes it possible to do an initial-placement algorithm that is better than the simple_placer, but right now the python runtime still maintains shape information that is not propagated to C++.  I think we need to wait for the control flow ops to be implemented via the C-API before we could make any more progress on initial static shape inference -- I believe the C-API stuff is being worked on.  There are probably some other ways to start playing around with this though (e.g., via add_shapes argument to as_graph_def()), but nobody I know is working on that.

**Comment by bhack** [comment]
I don't really know how you are solving this in production or managed cloud :) I know that you cannot disclose it but I don't think that you are handling manual placement.

**Comment by benoitsteiner** [comment]
We are now able to predict memory usage and performance for each operation of a TensorFlow graph. With this data we can start the actual work on optimizing TensorFlow models. We're completing work on a first optimizer that improves the performance of several models by about 20%, and we've also started to find a good device placement in parallel.
We'll start open sourcing this work as it matures. It's hard to predict how quickly research projects will succeed so unfortunately I don't have a roadmap to share.

**Comment by bhack** [comment]
@benoitsteiner Thank you for the update. We can live this open to track relevant progress.

**Comment by vrv** [comment]
Probably through hacks and common deployment patterns. :P

**Comment by bhack** [comment]
@vrv I've seen your deleted comment. Another hypothesis are a secret army of manual [Oompa-Loompa](http://roalddahl.wikia.com/wiki/Oompa_Loompas) device placers ;)

**Comment by vrv** [comment]
Indeed, Oompa-Loompa's are not simple people.

(My deleted comment was in response to "I don't really know how you are solving this in production" to which I responded "hacks and common deployment scenarios".  I didn't see all the other comments after and it looked out of context ;)

**Comment by bhack** [comment]
:smile:

**Comment by alquraishi** [comment]
Since #4552 was closed and referenced here, I would just like to point out that there's a separate issue having nothing to do with distributed computing, and that is to prevent TF from scheduling too many concurrent operations at once such that it runs out of memory (on a single GPU). I have run into this problem many times where every single op can fit onto memory but because TF is overly aggressive about concurrency I end up not being able to run the model.

**Comment by yangjunpro** [comment]
May I know the progress on this issue? Auto placement is a really nice shot for TF guys to benefit its distributed execution model. We have tried to do some kind of alike work by mixing data-parallelism with model-parallelism to get a trade-off between computation and communication and see promising improvement against standard data-parallelism implementation, also we are trying to automate this work. So maybe there are some potential collaboration with this issue.

@yuefengz @benoitsteiner

**Comment by benoitsteiner** [comment]
We've made a lot of progress on the placement side, which will be detailed in our upcoming ICML paper. 

We've also automated data-parallelism in the  [grappler auto parallelizer](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/auto_parallel.cc). We've tested it successfully on several models and we'll start documenting it as soon as possible.

We're open to a collaboration. Is there any documentation or paper you can point us to to get started thinking about what a common solution could look like ?

**Comment by thinxer** [comment]
Got the name of the paper: *Device Placement Optimization with Reinforcement Learning*. Looking forward to the preprint!

**Comment by yangjunpro** [comment]
@benoitsteiner

Thanks for the nice progress.
Our work haven't been recorded into any paper since we are still on the way with it. 
As to the grappler auto parallelizer, we do have taken a look at its implementation. It is nice and somehow is what we want, although currently it only support naive BSP auto placement(Is my understanding correct?). My question is that why not to implement the grappler auto-parallelizer in python side, since in python side, there are more graph metadata and the processing overhead should be less than implementing it in c++ side.

As to the starting point of our solution, we are working on designing a cost model to evaluate the potential placement solution to help find a sub-optimal one since we think the auto-placement problem actually is a NP-hard problem, it is non-trivial to find the optimal solution. 
We have added hybrid-parallelism for AlexNet into python side(actually transform the placement problem into a graph re-written problem), which is a mixture of data-parallelism and model parallelism and the speed-up against standard implementation is significant. The implementation overhead actually is not quite large. And also we are trying to follow this strategy to make it as automatic as possible, also as general as possible. 

Not sure how far your device placement optimization work has gone.

As to your nice paper(Device Placement Optimization with Reinforcement Learning) to be published on ICML 2017, will you put it on arXiv before the conference? 

Thanks

**Comment by benoitsteiner** [comment]
Our [Device Placement Optimization with Reinforcement Learning](https://arxiv.org/abs/1706.04972) paper has just been published on arXiv.

**Comment by yangjunpro** [comment]
@benoitsteiner 

Thanks for the update, already taken a go-through of your nice paper and really appreciate your fancy RL-based idea illustrated in the paper.
It looks that so far the device placement optimization is still time consuming(around 20 hour with 80 GPUs), so will you integrate it into the TF master revision? Also may I know inside Google how often you will use this RL-based automatic placement solution to accelerate the model training process?

Thanks

**Comment by SanZhiYuan** [comment]
hi, why not take a little step forward,  like users should explicitly declare where the computation should happens (cpu, gpu, ps), but just with a 'virtual' device id, then tf automatically schedule the sub graphs to any free devices, or a preferred device list would be even better for migration between distrubuted, multi-gpus, single gpu and cpu only. And I think this would be much easier for us to try those training tf models in our own environment.
PS, I really don't want to try the distributed tf after I checked the tutorials, -,-

**Comment by slothkong** [comment]
I read the paper. I only wish there was a bit more detail on the RL training procedure. My I ask when would we be able to read some more documentation on this topic?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by bhack** [comment]
This issue could have some connection with https://github.com/tensorflow/k8s/issues/35 /cc @jlewi

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bhack** [comment]
How the device placement is handled for the new high level api like `estimators`. Expecially with `tf.estimator.train_and_evaluate` where users don't handle `model_fn` manually to make explicit device assignment.

**Comment by bhack** [comment]
This last topic could impact also https://github.com/tensorflow/k8s/issues/61

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by alextp** [comment]
Closing this issue as it's too broad. For the common case where graphs have one compute device and many parameter server devices we currently support automatic placement. For more complicated cases it's an open research problem how to do this.

**Comment by 3a78f807111** [comment]
@benoitsteiner    I have read your  Device Placement Optimization with Reinforcement Learning paper, but I have some questions about it. First, this paper has given the cost function and used adam optimizer to update parameters, what's the use of formula (3) in this paper? Second, how can I calculate the probability of placement P? Finally, is the code for this paper open source? Thank you.

**Comment by bhack** [comment]
https://ai.googleblog.com/2020/12/end-to-end-transferable-deep-rl-for.html

**Comment by bhack** [comment]
https://arxiv.org/abs/2201.12023

**Comment by Stonepia** [comment]
> https://arxiv.org/abs/2201.12023
Hello, could I ask if there is a plan of open source the code? I am curious about what your paper mentioned :

```
Since there are only less than 80 kinds of common primitive operators in DL graphs, we can manually enumerate the possible parallel algorithms for all primitive operators.
```
Is there any doc or code relating to this?
